# API 資料串接 - 以 Dcard API 實作範例

* 了解 Dcard API 使用方式與回傳內容
* 撰寫程式存取 API 且解析 JSON 格式資料


## 作業目標

* 請利用 API: https://www.dcard.tw/_api/forums/pet/posts?popular=true 回答下列問題：

1. 這個 API 一次會回傳幾筆資料？每一筆資料包含哪些欄位？
2. 取出每一筆資料的「標題」、「貼文時間」、「留言人數」、「按讚人數」
3. 計算熱門/非熱門文章的「平均留言人數」與「平均按讚人數」

In [33]:
# 1. 這個 API 一次會回傳幾筆資料？每一筆資料包含哪些欄位？    
# 這邊我可能先用json瀏覽器 jsonview來看
import requests
r = requests.get(' https://www.dcard.tw/_api/forums/pet/posts?popular=true')
response = r.text

import json
data = json.loads(response)
data

[{'id': 233267345,
  'title': '請大家不要隨便養兔兔！',
  'excerpt': '兔兔很可愛，但是請不要隨意養他，因為，———，1.看醫生很貴/一定要到兔專科看！便宜幾百貴的幾萬！，2.兔子是很敏感的/一不吃不喝立即有危險！，3.不要看太多兔子卡通/真正的食物不是紅蘿蔔！，4.兔子',
  'anonymousSchool': False,
  'anonymousDepartment': True,
  'pinned': False,
  'forumId': '7ee21581-1307-4ba9-b9dc-82028bdcca49',
  'replyId': None,
  'createdAt': '2020-03-15T08:50:39.336Z',
  'updatedAt': '2020-03-15T08:50:39.336Z',
  'commentCount': 102,
  'likeCount': 1254,
  'withNickname': False,
  'tags': [],
  'topics': ['兔子', '領養代替購買', '貓', '寵物', '兔'],
  'meta': {'layout': 'classic'},
  'forumName': '寵物',
  'forumAlias': 'pet',
  'gender': 'F',
  'school': '宏國德霖科技大學',
  'replyTitle': None,
  'mediaMeta': [{'id': '7defd06f-1b3b-4a6b-a270-725bbb2e8865',
    'url': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89cfa319eace/640.jpeg',
    'normalizedUrl': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89cfa319eace/640.jpeg',
    'thumbnail': 'https://megapx-assets.dcard.tw/images/5fd32e54

一次回傳30筆資料

In [32]:
c=0
for d in data:
    c=c+1
    print(c,d['title'] ,"\n")

1 請大家不要隨便養兔兔！ 

2 我爸的兄弟 

3 狗得了環狀病毒（文長） 

4 啊～我是貓啦 

5 天使 

6 購買跟領養 

7 我以為我養的是法鬥 原來是隻豬啊 

8 感情好到讓人羨慕 

9 我家的特大號壁虎 

10 沒錢別養寵物 

11 守宮其實很可愛！ 

12 （圖多）最近從朋友那領養了一隻貓，正式晉升為貓奴才了！每天都會被他笑死 

13 我是不是養錯龜了..... 

14 犬種介紹特輯13 ~ 小姑娘 可卡犬 

15 很憤怒的領養經驗😡 

16 新成員 (圖多） 

17 我叫阿秋 真的很秋 

18 通緝金門大學ㄉ女同學的狗 

19 療癒！石虎寶寶長大了😚 

20 彰化尋狗！彰師大附近！拜託大家幫幫忙了！ 

21 還有人養魚當寵物嗎 

22 舌頭給我收斂一點！ 

23 外拍初體驗 

24 蒐集貓咪の各種睡姿 

25 只有我覺得這樣很危險嗎⋯⋯ 

26 挑軟柿子是吧？ 

27 協尋鸚鵡 

28 咪咪自己開水喝 

29 專業手機架 

30 轉：3/11在台三線的重機騎士麻煩進來看看 



先用json瀏覽器 jsonview來看，第一筆資料包含這些欄位。

id: 233267345,
title: "請大家不要隨便養兔兔！",
excerpt: "兔兔很可愛，但是請不要隨意養他，因為，———，1.看醫生很貴/一定要到兔專科看！便宜幾百貴的幾萬！，2.兔子是很敏感的/一不吃不喝立即有危險！，3.不要看太多兔子卡通/真正的食物不是紅蘿蔔！，4.兔子",
anonymousSchool: false,
anonymousDepartment: true,
pinned: false,
forumId: "7ee21581-1307-4ba9-b9dc-82028bdcca49",
replyId: null,
createdAt: "2020-03-15T08:50:39.336Z",
updatedAt: "2020-03-15T08:50:39.336Z",
commentCount: 102,
likeCount: 1254,
withNickname: false,
tags: [ ],
topics: [],
meta: {},
forumName: "寵物",
forumAlias: "pet",
gender: "F",
school: "宏國德霖科技大學",
replyTitle: null,
mediaMeta: [],
reactions: [],
hidden: false,
customStyle: null,
isSuspiciousAccount: false,
layout: "classic",
withImages: true,
withVideos: false,
media: [],
reportReasonText: "",
postAvatar: ""

In [34]:
#全部印出來的話會是這樣
for d in data:
    c=c+1
    print(c,d['title'] ,"\n")
    print(d['id'] ,"\n")
    print(d['excerpt'] ,"\n")
    print(d['anonymousSchool'] ,"\n")
    print(d['anonymousDepartment'] ,"\n")
    print(d['pinned'] ,"\n")
    print(d['forumId'] ,"\n")
    print(d['replyId'] ,"\n")
    print(d['createdAt'] ,"\n")
    print(d['updatedAt'] ,"\n")
    print(d['commentCount'] ,"\n")
    print(d['likeCount'] ,"\n")
    print(d['withNickname'] ,"\n")
    print(d['tags'] ,"\n")
    print(d['topics'] ,"\n")
    print(d['meta'] ,"\n")
    print(d['forumName'] ,"\n")
    print(d['forumAlias'] ,"\n")
    print(d['gender'] ,"\n")
    print(d['school'] ,"\n")
    print(d['replyTitle'] ,"\n")
    print(d['mediaMeta'] ,"\n")
    print(d['reactions'] ,"\n")
    print(d['hidden'] ,"\n")
    print(d['customStyle'] ,"\n")
    print(d['isSuspiciousAccount'] ,"\n")
    print(d['layout'] ,"\n")
    print(d['withImages'] ,"\n")
    print(d['withVideos'] ,"\n")
    print(d['media'] ,"\n")
    print(d['reportReasonText'] ,"\n")
    print(d['postAvatar'] ,"\n")

31 請大家不要隨便養兔兔！ 

233267345 

兔兔很可愛，但是請不要隨意養他，因為，———，1.看醫生很貴/一定要到兔專科看！便宜幾百貴的幾萬！，2.兔子是很敏感的/一不吃不喝立即有危險！，3.不要看太多兔子卡通/真正的食物不是紅蘿蔔！，4.兔子 

False 

True 

False 

7ee21581-1307-4ba9-b9dc-82028bdcca49 

None 

2020-03-15T08:50:39.336Z 

2020-03-15T08:50:39.336Z 

102 

1254 

False 

[] 

['兔子', '領養代替購買', '貓', '寵物', '兔'] 

{'layout': 'classic'} 

寵物 

pet 

F 

宏國德霖科技大學 

None 

[{'id': '7defd06f-1b3b-4a6b-a270-725bbb2e8865', 'url': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89cfa319eace/640.jpeg', 'normalizedUrl': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89cfa319eace/640.jpeg', 'thumbnail': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89cfa319eace/640.jpeg', 'type': 'image/thumbnail', 'tags': ['ANNOTATED'], 'createdAt': '2020-03-15T08:50:39.336Z', 'updatedAt': '2020-03-15T08:50:39.336Z', 'width': 480, 'height': 640}, {'id': '7defd06f-1b3b-4a6b-a270-725bbb2e8865', 'url': 'https://megapx-assets.dcard.tw/images/5fd32e54-9089-4f05-9a1f-89

共有這些欄位

In [29]:
# 2. 取出每一筆資料的「標題」、「貼文時間」、「留言人數」、「按讚人數」
c=0
for d in data:
    c=c+1
    print(c,"標題:",d['title'] ,"\n")
    print("貼文時間:",d['createdAt'] ,"\n")
    print("留言人數:",d['commentCount'] ,"\n")
    print("按讚人數:",d['likeCount'] ,"\n")

1 標題: 請大家不要隨便養兔兔！ 

貼文時間: 2020-03-15T08:50:39.336Z 

留言人數: 93 

按讚人數: 1121 

2 標題: 我爸的兄弟 

貼文時間: 2020-03-15T07:15:09.251Z 

留言人數: 9 

按讚人數: 1003 

3 標題: 狗得了環狀病毒（文長） 

貼文時間: 2020-03-15T11:50:11.895Z 

留言人數: 28 

按讚人數: 677 

4 標題: 啊～我是貓啦 

貼文時間: 2020-03-15T10:10:55.945Z 

留言人數: 9 

按讚人數: 470 

5 標題: 天使 

貼文時間: 2020-03-15T09:42:18.812Z 

留言人數: 8 

按讚人數: 188 

6 標題: 購買跟領養 

貼文時間: 2020-03-15T07:37:42.641Z 

留言人數: 5 

按讚人數: 73 

7 標題: 我以為我養的是法鬥 原來是隻豬啊 

貼文時間: 2020-03-16T02:54:35.422Z 

留言人數: 1 

按讚人數: 42 

8 標題: 感情好到讓人羨慕 

貼文時間: 2020-03-15T10:05:05.271Z 

留言人數: 4 

按讚人數: 42 

9 標題: 我家的特大號壁虎 

貼文時間: 2020-03-16T05:38:57.393Z 

留言人數: 3 

按讚人數: 28 

10 標題: 沒錢別養寵物 

貼文時間: 2020-03-15T09:05:46.214Z 

留言人數: 6 

按讚人數: 28 

11 標題: 失去貓咪後失去重心 

貼文時間: 2020-03-15T03:16:12.236Z 

留言人數: 9 

按讚人數: 27 

12 標題: 守宮其實很可愛！ 

貼文時間: 2020-03-15T18:14:56.035Z 

留言人數: 13 

按讚人數: 20 

13 標題: （圖多）最近從朋友那領養了一隻貓，正式晉升為貓奴才了！每天都會被他笑死 

貼文時間: 2020-03-15T14:44:32.318Z 

留言人數: 6 

按讚人數: 19 

14 標題: 我是不是養錯龜了..... 

貼文時間: 2020-03-

In [39]:

#先計算總共平均數
commentCount = 0
likeCount =0
for d in data:
    commentCount= commentCount + d['commentCount']
    likeCount =likeCount + d['likeCount']
    
print(commentCount/30 ,",", likeCount/30)


8.466666666666667 , 145.96666666666667


In [44]:
# 3. 計算熱門/非熱門文章的「平均留言人數」與「平均按讚人數」
# 假設定義 平均按讚人數超過100算熱門文章
          
hot_commentCount = 0
unhot_commentCount = 0

hot_likeCount =0
unhot_likeCount =0

hot_PostCount = 0
unhot_PostCount = 0
for d in data:
    if d['likeCount'] >=100 :
        hot_PostCount = hot_PostCount+1
        hot_commentCount= hot_commentCount + d['commentCount']
        hot_likeCount= hot_likeCount + d['likeCount']
    else:
        unhot_PostCount = unhot_PostCount+1
        unhot_commentCount=unhot_commentCount+d['commentCount']
        unhot_likeCount = unhot_commentCount +d['likeCount']

print("熱門文章平均留言人數:",hot_commentCount/hot_PostCount)
print("熱門文章平均按讚人數:",hot_likeCount/hot_PostCount)
print("非熱門文章平均留言人數:",unhot_commentCount/unhot_PostCount)
print("非熱門文章平均按讚人數:",unhot_likeCount/unhot_PostCount)


熱門文章平均留言人數: 31.6
熱門文章平均按讚人數: 780.6
非熱門文章平均留言人數: 3.84
非熱門文章平均按讚人數: 4.12
